### PR TITLE
Prevent pulling when teleporting

### LINKED ISSUE #33243: Ninjas can teleport with dragged objects
## Description

Ninjas will often cheese the RD server or console objectives by teleporting into space with it. They may also stun someone (#33242) and teleport into space to kill them.

**Reproduction**

1. Drag object
2. Dash

**Additional context**
I recall a similar issue when activating a scram implant while being dragged

### PR DESCRIPTION
**Pull request by thetolbean**
## About the PR
If a user attempts to teleport, either through the space ninja's katana or a scram implant, they will drop said item before teleporting

## Why / Balance
Fix #33243 and prevent teleports from being extended to pulled objects

## Technical details
Adds a check to both `OnScramImplant` & `OnDash` that check if the user is pulling an item, and if so, attempt to cancel the pull before the teleport

## Media

https://github.com/user-attachments/assets/ed8b0c71-0beb-4b37-abca-9c88b9c691e4

## Requirements

- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.

**Changelog**

:cl:
- tweak: Teleporting or dashing now causes you to stop pulling objects.

### CODE REVIEW

**Review by slarticodefast — APPROVED** [review]

**Comment by slarticodefast** [comment]
Tested and works as intended, thanks for the PR!

> I did say the issue was present with scram implants, but I also heard it was funny to scram into space with a secoff that was dragging you earlier, so it might be worth leaving that in considering its unpredictability

While I agree with you that it would be funnier to keep, this has too many potential edge cases where it could cause bugs with pulling.

**Comment by MilenVolf** [comment]
Does it also prevent pulling if user or targes is being teleported? For example, by bluespace anomaly

**Comment by slarticodefast** [comment]
> Does it also prevent pulling if user or target is being teleported? For example, by bluespace anomaly

The bluespace anomaly is a separate system, but it likely has the same problem.

**Comment by thetolbean** [comment]
> Does it also prevent pulling if user or target is being teleported? For example, by bluespace anomaly

Yes, as Slarti said the bluespace system is separate
With being pulled however, the scram implant already had a check in place to escape from being pulled. The dash does not, so I will be adding that shortly

**Comment by NoElkaTheGod** [comment]
It was kinda funny though. But removing this is probably for the better.

File: Content.Shared/Ninja/Systems/DashAbilitySystem.cs [hunk 1]
```diff
@@ -80,6 +85,14 @@ private void OnDash(Entity<DashAbilityComponent> ent, ref DashEvent args)
             return;
         }
 
+        // Check if the user is BEING pulled, and escape if so
+        if (TryComp<PullableComponent>(user, out var pull) && _pullingSystem.IsPulled(user, pull))
+            _pullingSystem.TryStopPull(user, pull);
+
+        // Check if the user is pulling anything, and drop it if so
+        if (TryComp(user, out PullerComponent? puller) && TryComp(puller.Pulling, out PullableComponent? pullable))
```
Comment (slarticodefast) [inline Content.Shared/Ninja/Systems/DashAbilitySystem.cs:93]:
```suggestion
        if (TryComp<PullerComponent>(user, out var puller) && TryComp<PullableComponent>(puller.Pulling, out var pullable))
```
maybe make it a little more consistent for better readability

File: Content.Server/Implants/SubdermalImplantSystem.cs [hunk 2]
```diff
@@ -109,6 +109,10 @@ private void OnScramImplant(EntityUid uid, SubdermalImplantComponent component,
         if (TryComp<PullableComponent>(ent, out var pull) && _pullingSystem.IsPulled(ent, pull))
             _pullingSystem.TryStopPull(ent, pull);
 
+        // Check if the user is pulling anything, and drop it if so
+        if (TryComp(ent, out PullerComponent? puller) && TryComp(puller.Pulling, out PullableComponent? pullable))
```
Comment (slarticodefast) [inline Content.Server/Implants/SubdermalImplantSystem.cs:113]:
```suggestion
        if (TryComp<PullerComponent>(ent, out var puller) && TryComp<PullableComponent>(puller.Pulling, out var pullable))
```

**Review by slarticodefast — CHANGES_REQUESTED** [review]

**Review by slarticodefast — APPROVED** [review]

**Comment by slarticodefast** [comment]
Thanks!

**Comment by thetolbean** [comment]
And I did perform testing on the bluespace scenarios, both the linked lockers and bluespace anom.
Both cancel your pulls if you are dragging something.
The items come along in the locker if they are also closed in, but you are not pulling them when you leave the other locker

**Comment by thetolbean** [comment]
> It was kinda funny though. But removing this is probably for the better.

Yeah, I agree. There's nothing like stealing an entire computer (or person) in the blink of an eye
Although it definitely needs to be removed for balancing

**Comment by chromiumboy** [comment]
My thoughts are that the pulling system should listen for when a puller's transform get updated, and if the displacement is outside a certain tolerance, the pull is canceled. That way all teleport cases would be covered

**Pull request merged**